f1 metric - micro/macro/weighted

SSaishruthi opened the pull request:
This PR contains a script for adding f1-macro metric.

Colab notebook: https://colab.research.google.com/drive/1LJ1yb8cUgisNP3ETfMLRYPQEyJivzDv-

SSaishruthi commented:
Please let me know what will be my next steps. To do as far as I know will be to add a test for the same.

Squadrick commented on tensorflow_addons/metrics/f1_macro.py:
```diff
@@ -0,0 +1,63 @@
+import tensorflow as tf
```
Add license and future import like [here](https://github.com/AakashKumarNain/addons/blob/0c31db81ac4d69d63e9fcf0a79aae9feed8e2176/tensorflow_addons/metrics/cohens_kappa.py#L1)

Squadrick commented on tensorflow_addons/metrics/f1_macro.py:
```diff
@@ -0,0 +1,63 @@
+    def result(self):
+        p_sum = tf.cast(self.true_positives_col + self.false_positives_col,
+                        tf.float32)
+        precision_macro = tf.cast(tf.compat.v1.div_no_nan(
```
Replace: `tf.compat.v1.div_no_nan` -> [`tf.math.divide_no_nan`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/math/divide_no_nan)

Squadrick commented on tensorflow_addons/metrics/f1_macro.py:
```diff
@@ -0,0 +1,63 @@
+
+        r_sum = tf.cast(self.true_positives_col + self.false_negatives_col,
+                        tf.float32)
+        recall_macro = tf.cast(tf.compat.v1.div_no_nan(
```
ditto

Squadrick commented on tensorflow_addons/metrics/f1_macro.py:
```diff
@@ -0,0 +1,63 @@
+
+        mul_value = 2 * precision_macro * recall_macro
+        add_value = precision_macro + recall_macro
+        f1_macro = tf.cast(tf.compat.v1.div_no_nan(mul_value, add_value),
```
ditto

SSaishruthi commented:
Thanks @Squadrick 
Will update soon

SSaishruthi commented:
@Squadrick 
Updated with F1 micro, macro and weighted
Test colab notebook: https://colab.research.google.com/drive/1qSq0SsYkPqjdKUgM1RM4kKM67X75ocFj

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+    """
+    Calculates F1 micro score
+    """
+    def __init__(self, name='f1-score'):
```
Nit: f1-score -> f1_score

WindQAQ replied:
I think `F1Micro` should be a better Python class name.

BTW, my thought on different average scheme for f1 score:
```python
class F1Score(Metric):
    def __init__(self, name="f1_score", average="micro"):
        # average could be one of "micro", "macro" or "weighted"
```
This would be much close to sklearn API.
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html

I'm also fine with seperated `F1Micro`, `F1Macro` and `F1Weighted` though. What do you free about this @Squadrick.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+    """
+    def __init__(self, name='f1-score'):
+        super(F1_micro, self).__init__(name=name)
+        self.true_positives = self.add_weight('TP-class', shape=[],
```
Nit: TP-class -> TP_class

WindQAQ replied:
A suggestion here: maybe "true_positives" is a better name to match core TF.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L1160

Squadrick replied:
Yup, it's better we match core TF.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        self.true_positives = self.add_weight('TP-class', shape=[],
+                                              initializer='zeros',
+                                              dtype=tf.float32)
+        self.false_positives = self.add_weight('FP-class', shape=[],
```
Nit: FP-class -> FP_class

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        self.false_positives = self.add_weight('FP-class', shape=[],
+                                               initializer='zeros',
+                                               dtype=tf.float32)
+        self.false_negatives = self.add_weight('FN-class', shape=[],
```
Nit: FN-class -> FN_class

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        p_sum = tf.cast(self.true_positives + self.false_positives,
+                        tf.float32)
+        # precision calculation
+        precision_micro = tf.cast(tf.math.divide_no_nan(
```
No need to cast. `p_sum` and `self.true_positives` are both `tf.float32`

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        r_sum = tf.cast(self.true_positives + self.false_negatives,
+                        tf.float32)
+        # recall calculation
+        recall_micro = tf.cast(tf.math.divide_no_nan(
```
ditto

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+
+        mul_value = 2 * precision_micro * recall_micro
+        add_value = precision_micro + recall_micro
+        f1_micro = tf.cast(tf.math.divide_no_nan(mul_value, add_value),
```
No need to cast. `mul_value` and `add_value` are both `tf.float32`.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+    Calculates F1 macro or weighted based on the user's choice
+    """
+    def __init__(self, num_classes, average,
+                 name='f1-macro-and-weighted-score'):
```
Nit: f1-macro-and-weighted-score -> f1_macro_and_weighted_score

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        super(F1_macro_and_weighted, self).__init__(name=name)
+        self.num_classes = num_classes
+        self.average = average
+        self.true_positives_col = self.add_weight('TP-class',
```
Nit: TP-class -> TP_class

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+                                                  shape=[self.num_classes],
+                                                  initializer='zeros',
+                                                  dtype=tf.float32)
+        self.false_positives_col = self.add_weight('FP-class',
```
ditto

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+                                                   shape=[self.num_classes],
+                                                   initializer='zeros',
+                                                   dtype=tf.float32)
+        self.false_negatives_col = self.add_weight('FN-class',
```
ditto

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+                                                   shape=[self.num_classes],
+                                                   initializer='zeros',
+                                                   dtype=tf.float32)
+        self.weights_intermediate = self.add_weight('weights-int-f1',
```
ditto

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        p_sum = tf.cast(self.true_positives_col + self.false_positives_col,
+                        tf.float32)
+        # calculate precision
+        precision_macro = tf.cast(tf.math.divide_no_nan(
```
Remove cast.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+        r_sum = tf.cast(self.true_positives_col + self.false_negatives_col,
+                        tf.float32)
+        # calculate recall
+        recall_macro = tf.cast(tf.math.divide_no_nan(
```
Remove cast.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,166 @@
+
+        mul_value = 2 * precision_macro * recall_macro
+        add_value = precision_macro + recall_macro
+        f1_macro_int = tf.cast(tf.math.divide_no_nan(mul_value, add_value),
```
Remove cast.

SSaishruthi commented:
@Squadrick 
Updated scripts. Please let me know if there is any change to be made.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -0,0 +1,177 @@
+        f1_score = tf.reduce_mean(f1_macro_int)
+        # condition for weighted f1 score
+        if self.average == 'weighted':
+            f1_int_weights = tf.cast(tf.math.divide_no_nan(
```
Remove cast, not needed.

SSaishruthi commented:
Hi @Squadrick @WindQAQ 

I have updated code to calculate F1 scores under one class.
Now, we have four options:

1. Micro
2. Macro
3. Weighted
4. None (This will produce F1 scores for each class)

Colab: https://colab.research.google.com/drive/1qSq0SsYkPqjdKUgM1RM4kKM67X75ocFj#scrollTo=hZSxoW4eYqYW

I ran code format to check scripts. I encountered an installation error in `clang` and removed that step because I noticed that is only for C++.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -23,124 +23,85 @@
-        self.true_positives_col = self.add_weight('true_positives',
+            if self.average == 'micro':
+                self.axis = None
+            if self.average == 'macro' or self.average == 'weighted':
```
Use `else` here, no need for an `if`.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -23,124 +23,85 @@
+        super(F1Score, self).__init__(name=name)
         self.num_classes = num_classes
-        if average not in (None, 'macro', 'weighted'):
+        if average not in (None, 'micro', 'macro', 'weighted'):
```
What happens when `average` is set to `None`?

Squadrick replied:
Because it passes the check, but the calculation doesn't consider a case when `average=None`.

SSaishruthi replied:
When the average is set to `None`, F1-score is calculated for each class. This condition is handled in https://github.com/SSaishruthi/addons/blob/9348474f998e3bf0438f541832d98e31ea3fc2db/tensorflow_addons/metrics/f1_scores.py#L173

I got reference for `None` from sklearn. Computed is similar to `Macro` but mean will not be computed.

SSaishruthi replied:
I did not add `None` condition before. Added that today. So we have four conditions now:

Micro/ Macro / weighted /None

SSaishruthi replied:
Also, added a test condition for this.

Squadrick commented:
@SSaishruthi Can you pull from `master` and update this PR?

SSaishruthi commented:
@Squadrick 

I have updated the branch and resolved conflicts.

Squadrick commented on tensorflow_addons/metrics/f1_test.py:
```diff
@@ -0,0 +1,125 @@
+        self.evaluate(tf.compat.v1.variables_initializer(f1_obj2.variables))
+        return f1_obj, f1_obj1, f1_obj2
+
+    def initialize_vars_none(self):
```
Combine `f1_none` into `initialize_vars`. It's better than having a separate method that does the exact same thing.

Squadrick commented on tensorflow_addons/metrics/f1_test.py:
```diff
@@ -0,0 +1,125 @@
+    def initialize_vars(self):
+        f1_obj = F1Score(num_classes=3, average='micro')
+        f1_obj1 = F1Score(num_classes=3, average='macro')
+        f1_obj2 = F1Score(num_classes=3, average='weighted')
```
Nit: Rename `f1_obj` -> `f1_micro`, `f1_obj1` -> `f1_macro`, `f1_obj2` -> `f1_weighted`.

Squadrick replied:
Do the renaming through out the file.

Squadrick replied:
We can merge once as soon as this is done.

Squadrick commented on tensorflow_addons/metrics/f1_test.py:
```diff
@@ -0,0 +1,125 @@
+        self.evaluate(update_op2)
+        self.evaluate(update_op3)
+
+    def update_obj_states_none(self, f1_none, actuals, preds):
```
Refactor to include this in `update_obj_states`.

Squadrick commented on tensorflow_addons/metrics/f1_test.py:
```diff
@@ -0,0 +1,125 @@
+        self.check_results(f1_obj1, 0.6555555)
+        self.check_results(f1_obj2, 0.6777777)
+
+    def test_f1_none_score(self):
```
Rather make making a separate test case for `average=None`, include it in the other `test_f1_*` methods. So we can check for `perfect`, `random` and `worst` for the `None` case.

SSaishruthi commented:
@Squadrick 

Inputs to `None` condition should be multiclass but not multilabel. That's why I created a separate function for `None` as they require a different set of inputs. Can we have it like this or if you can suggest me an alternative that would be great.

Thanks

Squadrick commented:
@SSaishruthi Oh, my bad. In that case, leave it as is.

Squadrick commented:
Triggering the test cases.

SSaishruthi commented:
@Squadrick updated the scripts. Can you please trigger the test again?
